synapse/handlers/room.py

```diff
@@ -1025,14 +1028,21 @@ async def send(etype: str, content: JsonDict, **kwargs: Any) -> int:
             # Allow these events to be sent even if the user is shadow-banned to
             # allow the room creation to complete.
             (
-                _,
+                sent_event,
                 last_stream_id,
             ) = await self.event_creation_handler.create_and_send_nonmember_event(
                 creator,
                 event,
                 ratelimit=False,
                 ignore_shadow_ban=True,
+                prev_event_ids=[sent_event_ids[-1]] if sent_event_ids else None,
+                state_event_ids=state_event_ids.copy() if state_event_ids else None,
```

Comment:
I was somewhat paranoid about this, but I think this is fine. Here's my reasoning.

If I `send()` two events with the same (type, state_key) pair, they'll both appear in the `state_event_ids` variable. But the `state_event_ids` argument---representing the state of the room---ought to contain just the latest of the two events. 

(It's not completely clear to me if we will ever `send()` the same (type, state_key) pair twice. I think it depends on what we are given in the `initial_state`?)

Digging a little deeper, the state_event_ids propagate through a few layers:

- synapse.handlers.message.EventCreationHandler.create_and_send_nonmember_event
- synapse.handlers.message.EventCreationHandler.create_event
- synapse.handlers.message.EventCreationHandler.create_new_client_event

to here:

https://github.com/matrix-org/synapse/blob/92103cb2c8b8bff6b522a7bfa8a3a776b4821b11/synapse/handlers/message.py#L1034-L1053

That does the deduplication, building a state map where the most recent event with a given (type_state_key) pair "wins".

Reply:
Ah I was assuming here that room creation would never create duplicate (type, state_key) pairs, but this isn't quite true with the user supplied initial state. Looks like this is OK though as handled further down?

Reply:
That's my understanding, yeah!